### PR TITLE
Feature asset inventory changes

### LINKED ISSUE #6348: Asset Management - Inventory Changes
Our goal is to use the current stock lots as assets.  In order to do this, we need to add some data to the 'inventory' table and add some related functionality.

### Database Changes
- add: ‘is_asset’ to inventory table (boolean, default to false)
- add: ‘external_reference_number’ to inventory table (TEXT, defaults to NULL).  
- To be used to refer to an external reference such as PROSANI inventory ID.
- add: ‘manufacturer_brand’ to inventory table (marque) field (TEXT, defaults to NULL)
- add: ‘manufacturer_model’ field to inventory table (TEXT, defaults to NULL)
- add: ‘equipment_type_id’ to inventory table (reference into an equipment type table)

**Create ‘equipment_type’ table including these fields:**
- id
- label (TEXT, raw string or i18n token)
- description (TEXT)
- is_predefined (if label is i18n token)

**Notes**
- ‘inventory group’ is not quite suitable for equipment types.
- The equipment_type ‘label’ field will be the i18n token for predefined entries but raw text in case entered manually at a site.  Note that something like this is done in the cost_center_allocation_basis table and related code.

### Interface changes
- In the Inventory registry, add the new fields to the grid.  They should be hidden by default.
- The form to edit inventory items should allow editing these new fields.
- Add a simple editor for the equipment_type table entries.  It should not be possible to edit the label for equipment_types with predefined tokens.
- Add the ability to export the inventory registry as XLSX.

### PR DESCRIPTION
**Pull request by lomamech**
- Add property _is_asset, external_reference_number, manufacturer_brand, manufacturer_model_, in table **inventory**
- Use table **inventory_type** instead creat table **equipement_type**
- Complete integration test
- And constaint in form of inventory creation, allow reference number,
  manufacturer brand and manufacturer model only of Asset

closes #6348

### CODE REVIEW

File: client/src/modules/inventory/configuration/types/modals/actions.modal.js
Comment (jmcameron) [inline client/src/modules/inventory/configuration/types/modals/actions.modal.js:70]: 👍

File: client/src/modules/inventory/configuration/types/types.html
Comment (jmcameron) [inline client/src/modules/inventory/configuration/types/types.html:44]: Shouldn't there be a 'translate' if it is predefined?

File: client/src/modules/inventory/configuration/types/types.js
Comment (jmcameron) [inline client/src/modules/inventory/configuration/types/types.js:78]: Ok.  Ignore my last comment.

File: client/src/modules/inventory/list/modals/actions.tmpl.html
Comment (jmcameron) [inline client/src/modules/inventory/list/modals/actions.tmpl.html:54]: Nice!  I like that you only show these asset-related fields if it is an asset.

**Review by jmcameron — COMMENTED** [review]
When I edit an inventory item (via its action menu 'Edit' link) and I save the modifications, the Inventory registry loses my columns definitions.   The columns definitions should be cached.

Regarding whether 'inventory_type' is okay to use rather than adding a new 'equipment_type' filed.  I think re-using 'inventory_type' is fine.  So I do not object to using 'inventory_type' for this purpose.  However, we should take another look at our current predefined items for this field:  Article, Assembly, Service (IMCK adds 'Honoraire mediccin', and 'SOJA').  I'm not sure how these are being used, put we probably need to add explanations about this change when we install a release at production sites with this change.

I think that VARCHAR(30) for inventory_type.text may be too small.  Since it can now contain a i18n token, it could easily be longer.  So I suggest increasing its size to VARCHAR(200) or 'text' (for reference see cost_center_allocation_basis).

Although the new 'external_reference_number' field for 'inventory' was intended for "External" references, it could potentially be used for internal references.  So I suggest changing the 'external_reference_number' field name to 'reference_number'.  I notice that the GUI never mentions 'external', so this seems like a simple change.

Otherwise it looks good.  I reviewed the code.   And I tested it with bhima_test and everything seemed to work okay.   I did not test the migration scripts on a production dataset.  

After making these changes, please ask me for a new review and I'll check the migration part.

File: client/src/modules/inventory/configuration/types/types.html
Comment (jmcameron) [reply client/src/modules/inventory/configuration/types/types.html:44]: I see that you took care of this later

**Comment by jmcameron** [comment]
I checked 'Stock Lots' after finishing the review.  Shouldn't the new Inventory fields be available in the Stock Lots registry (via Menu > Columns)?

**Comment by lomamech** [comment]
> When I edit an inventory item (via its action menu 'Edit' link) and I save the modifications, the Inventory registry loses my columns definitions. The columns definitions should be cached.

To keep the columns visible, click on the menu then on the Column button
image

Then you have to choose the columns to display
image

And to save the configuration you just have to click on the **Save Grid Configuration** button
image

**Comment by lomamech** [comment]
@jmcameron i need a new review, it took me longer because I had also anticipated with the issue #6349 **_Asset Management - Stock Lot changes_**

**Comment by jmcameron** [comment]
> And to save the configuration you just have to click on the **Save Grid Configuration** button !

Okay.  I tested this and it works fine.  However, I still wonder why we need the "Save Grid Configuration" function instead of caching any changes to the grid immediately.   I think this is a broader "usability" issue and should not slow down this PR.

**Comment by lomamech** [comment]
> > And to save the configuration you just have to click on the **Save Grid Configuration** button !
> 
> Okay. I tested this and it works fine. However, I still wonder why we need the "Save Grid Configuration" function instead of caching any changes to the grid immediately. I think this is a broader "usability" issue and should not slow down this PR.

This functionality exists on all the grids of the BHIMA system and each user can save the grid, the size, the visibility and the positioning of the columns

**Review by jmcameron — COMMENTED** [review]
I tried doing the migration procedure and ran into several problems:

- In the migration.sql, file, the change of inventory_type.text cannot include the 'AFTER' clause.  That makes sense because it is modifying an existing field, not added a new field.
- The definition of inventory_type.text needs to be updated in schema.sql (to VARCHAR(200) like the migration change).
- In the migration file (line 57), 'external_reference_number' needs to be changed to 'reference_number'

- The definitions in admin.sql of zRecalculateCostCenterAggregates() and zRecomputeLotBarcodes() need to be preceded by DROP PROCEDURE IF EXISTS <name>

Please make these updates and I'll try the migration procedure again.

File: server/models/admin.sql
Comment (jmcameron) [inline server/models/admin.sql:268]: Sorry, my mistake.  This addition was not necessary.  This statement is already in the SQL files somewhere.  When you create the migration file, you can find it.

File: server/models/migrations/next/migrate.sql
Comment (jmcameron) [inline server/models/migrations/next/migrate.sql:57]: 👍

**Review by jmcameron — COMMENTED** [review]
I tested migrating an production dataset with this PR and it worked okay.   I also tested displaying and editing asset-related data in the inventory registry and that worked fine.

Before releasing I suggest eliminating the extra DROP PROCEDURE (see the SQL code comments) and adding a DROP PROCEDURE fro zRecomputeLotBarcodes() in admin.sql.

I think this PR is ready to go, but I encourage getting an review by @jniles or @mbayopanda before merging.

File: client/src/modules/inventory/list/modals/asset.cell.tmpl.html
Comment (jniles) [inline client/src/modules/inventory/list/modals/asset.cell.tmpl.html:3]: You could just do this with a `ng-class={}` directive on a single `i`.  I don't think this has any performance penalty though.

File: client/src/modules/inventory/list/templates/reference_number.cell.html
Comment (jniles) [inline client/src/modules/inventory/list/templates/reference_number.cell.html:3]: Why does this need to be a template?  You can do this with just a cellStyle, right?

File: client/src/modules/inventory/list/list.html
Comment (jniles) [inline client/src/modules/inventory/list/list.html:91]: :+1:

**Review by jniles — APPROVED** [review]
This looks good to me.  Thanks @lomamech 

bors r+

**Pull request merged**
